### PR TITLE
Make `simplifyUnusedExpr` non-recursive

### LINKED ISSUE #5398: Transpiler stack overflow when loading `aws-csdk-lib`
### What version of Bun is running?

- 1.0.1
- 1.0.2
- 1.0.3

### What platform is your computer?

- Linux 5.4.254-175.358.amzn2int.x86_64 x86_64 x86_64
- Linux 5.10.192-182.736.amzn2.x86_64 x86_64 x86_64 (last Amazon AL2 EC2 AMI)

NOTE: This bug is NOT reproducible on:
- Linux 4.15.0-20-generic x86_64 x86_64
- Linux 5.4.0-1103-aws x86_64 x86_64 (ami-0d7fd770997502624)
- Linux 5.14.21-150500.53-default x86_64 x86_64
- Linux 5.15.0-58-generic x86_64 x86_64
- Linux 6.2.14-300.fc38.aarch64 aarch64 aarch64

### What steps can reproduce the bug?
```
mkdir test-bun-aws-cdk-lib-bug && \
cd test-bun-aws-cdk-lib-bug && \
bun add aws-cdk-lib constructs && \
echo 'import cdk from "aws-cdk-lib"; cdk;' > index.ts && \
bun run .
```

### What is the expected behavior?

exit with status 0

### What do you see instead?

[1]    30502 segmentation fault  bun run .

### Additional information

Would be great to understand what causes segmentation fault

### LINKED ISSUE #8021: Regression: CDK commands no longer working after bun 1.0.8
### What version of Bun is running?

1.0.21+837cbd60d

### What platform is your computer?

Darwin 23.2.0 arm64 arm (also happens in GitHub Actions using ubuntu-latest)

### What steps can reproduce the bug?

I've made a repo here https://github.com/codetalkio/bun-issue-cdk-repro with a minimal reproduction that sets up a new CDK project using AWS's [CDK Getting Started Guide](https://docs.aws.amazon.com/cdk/latest/guide/getting_started.html). 

First clone this repository and install dependencies:

```bash
$ git clone git@github.com:codetalkio/bun-issue-cdk-repro.git
$ cd bun-issue-cdk-repro/app
$ bun --revision
1.0.21+837cbd60d
$ bun install
...
```

The run `bun cdk synth` while still in the `bun-issue-cdk-repro/app` directory:

```bash
$ bun run cdk synth
cdk synth

Subprocess exited with error null
error: script "cdk" exited with code 1 (SIGHUP)
```

### What is the expected behavior?

CDK should start synthethizing the stack and give CloudFormation as output into the terminal:

```bash
$ cd bun-issue-cdk-repro/app
$ bun --revision
1.0.8+2a405f691e80725fe0b97b93afd3b8cfed13fa5f
$ bun run cdk synth
cdk synth
Resources:
  CDKMetadata:
    Type: AWS::CDK::Metadata
    Properties:
      Analytics: v2:deflate64:H4sIAAAAAAAA/zPSMzS00DNQTCwv1k1OydbNyUzSqw4uSUzO1nFOywtKLc4vLUpOBbGd8/NSMksy8/NqdfLyU1L1sor1y4wM9ECas4ozM3WLSvNKMnNT9YIgNADUwEUzWQAAAA==
    Metadata:
      aws:cdk:path: AppStack/CDKMetadata/Default
    Condition: CDKMetadataAvailable
Conditions:
  CDKMetadataAvailable:
    Fn::Or:
      - Fn::Or:
...
```

### What do you see instead?

Instead CDK will exit without giving any meaningful error at all:

```bash
$ bun run cdk synth
cdk synth

Subprocess exited with error null
error: script "cdk" exited with code 1 (SIGHUP)
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by paperclover**
### What does this PR do?

Fixes #5398
Fixes #8021

`simplifyUnusedExpr` is called on any expression which is not dead code, but has a value that is unused. For example

```ts
a; // a is unused
a = (b, c); // b is unused
a = [d] ? e : f; // [d] is unused because it truthy. f is dead code
```

It is not safe to make this a naive recursive function because some minifiers will build very large comma expressions, with thousands of parts. Bun represents this as a nested binary expression, and when it attempts to simplify the parts, it stack overflows.

TODO:
- [ ] Explain the method used here (maybe draw a diagram), it is kind of confusing and I will forget how this works in the future.
- [ ] Fix E.If handling, it has a subtle flaw.
- [ ] Performance testing. This is hot code as it is run on every expression statement. I have a feeling ArrayList is faster for this use-case. But I wanted to try SegmentedList because the idea of not re-allocating old items was cool.
- [ ] Tests!

### CODE REVIEW

**Comment by paperdave** [comment]
not ready to merge btw. i also see i edited launch.json and webkit module (aaaargh)

File: src/js_printer.zig [hunk 1]
```diff
@@ -778,6 +778,8 @@ fn NewPrinter(
 
         binary_expression_stack: std.ArrayList(BinaryExpressionVisitor) = undefined,
 
+        debug_allowed_to_print_missing: if (Environment.isDebug) bool else void = false,
```
Comment (Jarred-Sumner) [inline src/js_printer.zig]:
```suggestion
        debug_allowed_to_print_missing: if (Environment.isDebug) bool else void = if (Environment.isDebug) false else {},
```

**Comment by paperdave** [comment]
not ready.

image

**Comment by paperdave** [comment]
i will likely abandon this implementation for a more simple / narrow version of this. and then my misc refactors i'll just re-implement later.

**Comment by paperdave** [comment]
i close PR but intentionally am keeping the branch alive

**Pull request closed without merging**